src/basic/in-addr-util.c

```diff
@@ -571,3 +571,34 @@ int in_addr_prefix_from_string_auto(
         return 0;
 
 }
+
+void in_addr_data_hash_func(const void *p, struct siphash *state) {
+        const struct in_addr_data *a = p;
+
+        siphash24_compress(&a->family, sizeof(a->family), state);
+
+        if (a->family == AF_INET)
+                siphash24_compress(&a->address.in, sizeof(a->address.in), state);
+        else if (a->family == AF_INET6)
+                siphash24_compress(&a->address.in6, sizeof(a->address.in6), state);
+}
+
+int in_addr_data_compare_func(const void *a, const void *b) {
+        const struct in_addr_data *x = a, *y = b;
+
+        if (x->family != y->family)
+                return x->family - y->family;
```

Comment:
let's hope that `sa_family_t` is signed, because otherwise this is pretty ugly... 

I usually prefer doing explicit comparisons here and returning -1, 0, 1...